osm2streets/src/block.rs

```diff
@@ -168,123 +151,134 @@ impl Block {
     }
 }
 
+// Also returns the first intersection.
 fn walk_around(
     streets: &StreetNetwork,
-    start_road: RoadID,
-    clockwise: bool,
+    start_road_side: RoadSideID,
     sidewalks: bool,
-) -> Result<Vec<Step>> {
-    let start_i = streets.roads[&start_road].src_i;
-
-    let mut current_i = start_i;
-    let mut current_r = start_road;
+) -> Result<(Vec<RoadSideID>, IntersectionID)> {
+    let mut roads = Vec::new();
+
+    // We may start on a loop road on the "inner" direction
+    {
+        let start_r = &streets.roads[&start_road_side.road];
+        if start_r.src_i == start_r.dst_i {
+            let i = &streets.intersections[&start_r.src_i];
+            if !i.get_road_sides_sorted(streets).contains(&start_road_side) {
+                bail!("Starting on inner piece of a loop road");
+            }
+        }
+    }
 
-    let mut steps = vec![Step::Edge(current_r)];
+    // We need to track which side of the road we're at, but also which direction we're facing
+    let mut current_road_side = start_road_side;
+    // TODO This is convoluted. When we started with a lane, it was that lane's dst_i
+    let (orig_from_intersection, start_intersection) = {
+        let dir = start_road_side.get_outermost_lane(streets).dir;
+        let road = &streets.roads[&start_road_side.road];
+        if dir == Direction::Forward {
+            (road.src_i, road.dst_i)
+        } else {
+            (road.src_i, road.dst_i)
+        }
+    };
+    let mut current_intersection = start_intersection;
```

Comment:
I feel like all of this logic ought to be simpler if we trace tuples (road, left / right, forwards / backwards). I might try that in a future PR.